torch/_inductor/cudagraph_trees.py

```diff
@@ -634,47 +651,47 @@ def __init__(
                 assert out is None
                 self.outputs_metadata.append(None)
 
-        # initialized on first run
-        self.checkpointed_caching_state: Optional[AllocatorState] = None
+        self.graph.replay()
```

Comment:
So one big change is that, for the first run of cuda graph trees, previously you had spattered the record and the replay between init and run. Now, both the record/replay are in init, but you still have the optimization where you save `recorded_outputs` so that you can deliver them the first time run is called. Is this related to why you have to defer cudagraphify? Is there a reason why the logic has to be split across init/run like this?

Reply:
This all just seems really finely balanced. Why can't we just have a non-uniform API, one for the first record/replay, and the another for subsequent replays.

Reply:
Fixed